header.c

```diff
@@ -127,8 +168,30 @@ static int sam_hrecs_update_hashes(sam_hrecs_t *hrecs,
 
     /* Add to read-group hash? */
     if ((type>>8) == 'R' && (type&0xff) == 'G') {
-        sam_hrec_tag_t *tag;
+        sam_hrec_tag_t *tag = sam_hrecs_find_key(h_type, "ID", NULL);
         int nrg = hrecs->nrg;
+        khint_t k;
+
+        if (!tag) {
+            hts_log_error("Header includes @RG line with no ID: tag");
+            return -1;  // ID should be present, according to spec.
+        }
+        assert(tag->str && tag->len >= 3);
```

Comment:
Since the tag is parsed and checked here, there is no need to do it again at line 211. So the code from 205-228 could be replaced with 
```
        hrecs->rg[nrg].name = tag->str + 3;
        hrecs->rg[nrg].name_len = tag->len - 3;
        hrecs->rg[nrg].ty   = h_type;
        hrecs->rg[nrg].id   = nrg;

        int r;
        k = kh_put(m_s2i, hrecs->rg_hash, hrecs->rg[nrg].name, &r);
        if (-1 == r) return -1;
        kh_val(hrecs->rg_hash, k) = nrg;

Reply:
Yes, that does make it a bit simpler.